src/canvas/Canvas.ts

```diff
@@ -1460,193 +1458,126 @@ export class Canvas extends SelectableCanvas {
    * @private
    * @param {TPointerEvent} e Event object
    * @param {FabricObject} target
-   * @return {Boolean}
+   * @returns true if grouping occurred
    */
-  _shouldGroup(e: TPointerEvent, target?: FabricObject): boolean {
+  protected _handleGrouping(
+    e: TPointerEvent,
+    target?: FabricObject
+  ): this is AssertKeys<this, '_activeObject'> {
     const activeObject = this._activeObject;
-    // check if an active object exists on canvas and if the user is pressing the `selectionKey` while canvas supports multi selection.
-    return (
+    const isAS = isActiveSelection(activeObject);
+    if (
+      // check if an active object exists on canvas and if the user is pressing the `selectionKey` while canvas supports multi selection.
       !!activeObject &&
       this._isSelectionKeyPressed(e) &&
       this.selection &&
       // on top of that the user also has to hit a target that is selectable.
       !!target &&
       target.selectable &&
-      // if all pre-requisite pass, the target is either something different from the current
-      // activeObject or if an activeSelection already exists
-      // TODO at time of writing why `activeObject.type === 'activeSelection'` matter is unclear.
-      // is a very old condition uncertain if still valid.
-      (activeObject !== target || activeObject.type === 'activeSelection') &&
-      //  make sure `activeObject` and `target` aren't ancestors of each other
-      !target.isDescendantOf(activeObject) &&
-      !activeObject.isDescendantOf(target) &&
+      // group target and active object only if they are different objects
+      // else we try to find a subtarget of `ActiveSelection`
+      (activeObject !== target || isAS) &&
+      //  make sure `activeObject` and `target` aren't ancestors of each other in case `activeObject` is not `ActiveSelection`
+      // if it is then we want to remove `target` from it
+      (isAS ||
+        (!target.isDescendantOf(activeObject) &&
+          !activeObject.isDescendantOf(target))) &&
       //  target accepts selection
-      !target.onSelect({ e: e })
-    );
-  }
-
-  /**
-   * Handles active selection creation for user event
-   * @private
-   * @param {TPointerEvent} e Event object
-   * @param {FabricObject} target
-   */
-  _handleGrouping(e: TPointerEvent, target: FabricObject) {
-    let groupingTarget: FabricObject | undefined = target;
-    // Called always a shouldGroup, meaning that we can trust this._activeObject exists.
-    const activeObject = this._activeObject!;
-    // avoid multi select when shift click on a corner
-    if (activeObject.__corner) {
-      return;
-    }
-    if (groupingTarget === activeObject) {
-      // if it's a group, find target again, using activeGroup objects
-      groupingTarget = this.findTarget(e, true);
-      // if even object is not found or we are on activeObjectCorner, bail out
-      if (!groupingTarget || !groupingTarget.selectable) {
-        return;
-      }
-    }
```

Comment:
this is why `activeObject.type === 'activeSelection'`
https://github.com/fabricjs/fabric.js/blob/a39e01cf5ceae66809272051f77319345d8492d2/src/canvas/Canvas.ts#L1481-L1483

Reply:
moved it under `_handleGroup` => `isActiveSelection` block